app/code/community/Bolt/Boltpay/Helper/Api.php

```diff
@@ -165,121 +165,141 @@ public function verify_hook($payload, $hmacHeader)
+            $immutableQuote = Mage::getModel('sales/quote')->loadByIdWithoutStore($immutableQuoteId);
 
-        if(!$this->storeHasAllCartItems($immutableQuote)){
-            throw new Exception("Not all items are available in the requested quantities.");
-        }
+            // make sure this quote has not already processed
+            if ($immutableQuote->isEmpty()) {
+                throw new Exception("This order has already been processed by Magento.");
+            }
 
-        // check if the quotes matches, frontend only
-        if ( $sessionQuoteId && ($sessionQuoteId != $immutableQuote->getParentQuoteId()) ) {
-            throw new Exception("The Bolt order reference does not match the current cart ID. Cart ID: [$sessionQuoteId]"." Bolt Reference: [".$immutableQuote->getParentQuoteId()."]");
-        }
+            if(!$this->storeHasAllCartItems($immutableQuote)){
+                throw new Exception("Not all items are available in the requested quantities.");
+            }
 
-        // check if quote has already been used
-        if ( !$immutableQuote->getIsActive() ) {
-            throw new Exception("The quote has expired." );
-        }
+            // check if the quotes matches, frontend only
+            if ( $sessionQuoteId && ($sessionQuoteId != $immutableQuote->getParentQuoteId()) ) {
+                throw new Exception("The Bolt order reference does not match the current cart ID. Cart ID: [$sessionQuoteId]"." Bolt Reference: [".$immutableQuote->getParentQuoteId()."]");
+            }
 
-        // adding guest user email to order
-        if (!$immutableQuote->getCustomerEmail()) {
-            $email = $transaction->from_credit_card->billing_address->email_address;
-            $immutableQuote->setCustomerEmail($email);
-            $immutableQuote->save();
-        }
-        
-        // explicitly set quote belong to guest if customer id does not exist
-        $parentQuote = Mage::getModel('sales/quote')->loadByIdWithoutStore($immutableQuote->getParentQuoteId());
-        $immutableQuote
-            ->setCustomerIsGuest( (($parentQuote->getCustomerId()) ? false : true) )
-            ->save();
 
-        $immutableQuote->getShippingAddress()->setShouldIgnoreValidation(true)->save();
-        $immutableQuote->getBillingAddress()->setShouldIgnoreValidation(true)->save();
+            // check if quote has already been used
+            if ( !$immutableQuote->getIsActive() ) {
+                throw new Exception("The quote has expired." );
+            }
+          
+            // check if this order is currently being proccessed.  If so, throw exception
+            /* @var Mage_Sales_Model_Quote $parentQuote */
+            $parentQuote = Mage::getModel('sales/quote')->loadByIdWithoutStore($immutableQuote->getParentQuoteId());
+            if ($parentQuote->isEmpty() || !$parentQuote->getIsActive()) {
+                throw new Exception("The quote is currently being processed.");
```

Comment:
'The quote'. $immutableQuote->getParentQuoteId() .'is currently being processed for is already processed'